samplejava/src/main/java/com/example/chattutorial/ChannelActivity3.java

```diff
@@ -107,29 +120,43 @@ public void handleOnBackPressed() {
         });
 
         // Custom typing info header bar
-        TextView typingHeaderView = findViewById(R.id.typingHeaderView);
         String nobodyTyping = "nobody is typing";
-        typingHeaderView.setText(nobodyTyping);
-
-        // Obtain a ChannelController
-        ChatDomain.instance()
-                .getChannelController(cid)
-                .enqueue((result) -> {
-                    ChannelController channelController = result.data();
-
-                    // Observe typing users
-                    channelController.getTyping().observe(this, typingState -> {
-                        if (typingState.getUsers().isEmpty()) {
-                            typingHeaderView.setText(nobodyTyping);
-                        } else {
-                            List<String> userNames = new LinkedList<>();
-                            for (User user : typingState.getUsers()) {
-                                userNames.add(user.getName());
-                            }
-                            String typing = "typing: " + TextUtils.join(", ", userNames);
-                            typingHeaderView.setText(typing);
-                        }
-                    });
-                });
+        binding.typingHeaderView.setText(nobodyTyping);
+
+        // Observe typing events and update typing header depending on its state.
+        Flow<ChannelState> channelStateFlow = ChatClientExtensions.watchChannelAsState(ChatClient.instance(), cid, 30);
+        LiveData<TypingEvent> typingEventLiveData = Transformations.switchMap(
+                FlowLiveDataConversions.asLiveData(channelStateFlow),
+                channelState -> FlowLiveDataConversions.asLiveData(channelState.getTyping())
+        );
+
+        typingEventLiveData.observe(this, typingEvent -> {
+            String headerText;
+
+            if (typingEvent.getUsers().size() != 0) {
+                headerText = "typing: " + joinTypingUpdatesToUserNames(typingEvent);
+            } else {
+                headerText = nobodyTyping;
+            }
+
+            binding.typingHeaderView.setText(headerText);
+        });
+    }
+
+    // Helper method that transforms typing updates into a string
+    // containing typing member's names
+    @NonNull
+    private String joinTypingUpdatesToUserNames(@NonNull TypingEvent typingEvent) {
```

Comment:
We have to do this manually because `String.join()` is not compatible with our lowest supported API (21).